docs/operations/batch-processing.md

```diff
@@ -0,0 +1,136 @@
+# Batch Processing
+
+In order to answer queries, JanusGraph has to perform queries
+against the storage backend.
+In general, there are two ways of doing this:
+
+- Once data from the backend is needed, execute a backend
+  query and continue with the result.
+- Maintain a list of what data is needed.
+  Once the list reaches a certain size, execute a batched
+  backend query to fetch all of it at once.
+  
+The first option tends to be more responsive and consume less
+memory because the query can emit the first results very early
+without waiting for larger batches of queries to complete.
+This is also the option that JanusGraph uses by default.
+The second option can be configured in multiple ways which are
+explained below.
+
+## Configuration Options
+
+### No Batch Processing
+This is the default configuration of JanusGraph.
+In terms of graph traversals, the execution of queries is
+loosely coupled to the principle of Depth-First-Search.
+
+#### Use this configuration in use cases where for example ...
+- ... each query only accesses few vertices of the graph.
+- ... your application does not need the _full_ result set
+  immediately but rather requires a low latency for the first
+  results to arrive.
+
+#### Possible limitations
+- Traversing large neighborhoods can make the query slow.
+
+#### Steps to explicitly configure this option:
+- Ensure `query.batch` is set to `false`
+
+### Unrestricted Batch Processing
+
+Using this configuration, each step which traverses the Graph
+starting from a vertex (so e.g. `in()`, `outE()` and `values()`
+but not `inV()` or `otherV()`) becomes a blocking operator
```

Comment:
Thanks for explaining the current behavior! I am just curious: do you know if this "blocking operator" works for `valueMap()` step as well (when `query.batch` is true)? AFAIK, batch query does not work for `valueMap` step, but I don't know whether it follows BFS or DFS.

For example, consider below queries (assuming the index query hits Elasticsearch and edge store query hits Cassandra):

`g.V().has("someKeyIndexed", "value").values()` If I am understanding this doc section correctly, this query will fetch all results from ES and then fire a single (logical) batch query against Cassandra.

`g.V().has("someKeyIndexed", "value").valueMap()`  - not sure about this one. Does JanusGraph get all results from ES first and then fire multiple queries against Cassandra?

Reply:
In contrast to `values()` steps, `valueMap()` steps are not converted to `JanusGraphPropertyStep`s by `JanusGraphLocalQueryOptimizerStrategy`. Since the blocking behavior only applies to `JanusGraphPropertiesStep` and `JanusGraphVertexStep`, `valueMap()` is not affected by this.

Reply:
Thanks for clearing the doubt!